confgenerator/logging_modify_fields.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package confgenerator
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/filter"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+)
+
+type ModifyField struct {
+	// Source of value for this field
+	MoveFrom    string  `yaml:"move_from" validate:"omitempty,field,excluded_with=CopyFrom StaticValue"`
+	CopyFrom    string  `yaml:"copy_from" validate:"omitempty,field,excluded_with=MoveFrom StaticValue"`
+	StaticValue *string `yaml:"static_value" validate:"excluded_with=MoveFrom CopyFrom"`
+
+	// Name of variable with copied value
+	sourceVar string `yaml:"-"`
+
+	// Operations to perform
+	MapValues map[string]string `yaml:"map_values"`
+	Type      string            `yaml:"type" validate:"omitempty,oneof=integer float"`
+	OmitIf    string            `yaml:"omit_if" validate:"omitempty,filter"`
+}
+
+type LoggingProcessorModifyFields struct {
+	ConfigComponent `yaml:",inline"`
+	Fields          map[string]*ModifyField `yaml:"fields" validate:"dive,keys,field,endkeys"`
+}
+
+func (p LoggingProcessorModifyFields) Type() string {
+	return "modify_fields"
+}
+
+func (p LoggingProcessorModifyFields) Components(tag, uid string) []fluentbit.Component {
+	c, err := p.components(tag, uid)
+	if err != nil {
+		// It shouldn't be possible to get here if the input validation is working, so treat this as a code bug.
+		panic(err)
```

Comment:
These `panic` errors should never be thrown, ever. They indicate a bug in Ops Agent that we failed to catch through unit tests. All of this code only runs at config generation time, so yes, it can't ever panic due to the contents of a log entry.

There won't be any logging if fields are missing; it treats "missing" as a valid value, so trying to copy a missing field will just unset the new field, condition matching a missing field will evaluate false, etc. I think that's reasonable logic.

Reply:
Good to know.